### PR TITLE
Fix blittable array instead of copying

### PR DESCRIPTION
**Pull request by JeremyKuhne**
System.Drawing is doing a lot of unnecessary work. This is a start on trying to remove some of it. This change is meant to show the general pattern I intend to follow.

- Fixes the managed array and uses it directly instead of copying multiple times
- Adds some basic sanity tests to validate that we're getting the expected output
- Splits existing negative bezier tests into new file
- Adds a performance project

Output tests were generated and run before making the functional change.

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
Drawing Beziers before:

```
  System.Drawing.Common.Performance.Tests.dll                      | Metric                                        | Unit  | Iterations |    Average | STDEV.S |        Min |        Max
  :---------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| ----------:| -------:| ----------:| ----------:
   System.Drawing.Tests.Perf_Graphics_DrawBeziers.DrawBezier_Points | Duration                                      | msec  |    100     |     38.219 |   0.565 |     36.993 |     40.016
   System.Drawing.Tests.Perf_Graphics_DrawBeziers.DrawBezier_Points | Allocation Size on Benchmark Execution Thread | bytes |    100     | 1.680E+006 |   0.000 | 1.680E+006 | 1.680E+006  
```

After:

```
  System.Drawing.Common.Performance.Tests.dll                      | Metric                                        | Unit  | Iterations | Average | STDEV.S |    Min |    Max
  :---------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| -------:| -------:| ------:| ------:
   System.Drawing.Tests.Perf_Graphics_DrawBeziers.DrawBezier_Points | Duration                                      | msec  |    100     |  34.866 |   0.980 | 33.530 | 40.966
   System.Drawing.Tests.Perf_Graphics_DrawBeziers.DrawBezier_Points | Allocation Size on Benchmark Execution Thread | bytes |    100     |   0.000 |   0.000 |  0.000 |  0.000
```

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs [hunk 1]
```diff
@@ -860,29 +860,23 @@ public void DrawBezier(Pen pen, Point pt1, Point pt2, Point pt3, Point pt4)
         /// <summary>
         /// Draws a series of cubic Bezier curves from an array of points.
         /// </summary>
-        public void DrawBeziers(Pen pen, Point[] points)
+        public unsafe void DrawBeziers(Pen pen, Point[] points)
         {
             if (pen == null)
-            {
                 throw new ArgumentNullException(nameof(pen));
-            }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: I'm not sure why these braces are being removed.

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: Not needed per our guidelines, pulls the argument validation in a bit.

Comment (tannergooding) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: But also at the cost of making the code bigger in collapsed view 😉

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > But also at the cost of making the code bigger in collapsed view

I'd call it a plus that you don't have to expand for one line.

Comment (tannergooding) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: It really comes down to personal preference 😄 . For example, I find the native code in CoreCLR significantly harder to navigate because of all the one liners that don't have braces.

Comment (jkotas) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md : When editing files, keep new code and changes consistent with the style in the files

The prevalent style in this file is to use extra braces.

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > The prevalent style in this file is to use extra braces.

I think this is a bit of missing the forest for the trees in this case.

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: What's the forest and what are the trees here?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs [hunk 2]
```diff
@@ -860,29 +860,23 @@ public void DrawBezier(Pen pen, Point pt1, Point pt2, Point pt3, Point pt4)
...
-            IntPtr buf = SafeNativeMethods.Gdip.ConvertPointToMemory(points);
-            try
+            fixed (Point* p = points)
             {
-                int status = SafeNativeMethods.Gdip.GdipDrawBeziersI(new HandleRef(this, NativeGraphics), new HandleRef(pen, pen.NativePen),
-                                                      new HandleRef(this, buf), points.Length);
+                int status = SafeNativeMethods.Gdip.GdipDrawBeziersI(
+                    new HandleRef(this, NativeGraphics),
+                    new HandleRef(pen, pen.NativePen),
+                    new HandleRef(this, new IntPtr(p)), points.Length);
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: Do you plan to also address use of HandleRef?  That's an old pattern prior to SafeHandles.

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > Do you plan to also address use of HandleRef? 

I'm cognizant about it. I'm not sure that I'll deal with it myself, but I'll at least open a follow-up issue.

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs [hunk 3]
```diff
@@ -860,29 +860,23 @@ public void DrawBezier(Pen pen, Point pt1, Point pt2, Point pt3, Point pt4)
...
-            }
 
             if (points == null)
-            {
                 throw new ArgumentNullException(nameof(points));
-            }
 
-            IntPtr buf = SafeNativeMethods.Gdip.ConvertPointToMemory(points);
-            try
+            fixed (Point* p = points)
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: It looks like ConvertPointToMemory is doing some complicated marshaling.  That's all unnecessary, and Point is blittable and matching the corresponding native types in layout?

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: Yes, Point matches the native type and is blittable. I'm not sure how they stumbled into that loop originally.

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: The intent (forest) is making sure files aren't unreadable due to jarring differences in style. The tree is the specifics of the rule. This particular application of the rule is incredibly nitpicky and has no real benefit to readability.

**Comment by JeremyKuhne** [comment]
@dotnet-bot test OSX x64 Debug Build

Failed with #29940

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: The same pattern is used at about 60 places. It would be nice to get all of them fixed up.

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > The same pattern is used at about 60 places. 

Yep. Following up on it right now. Just wanted to get an initial change up to show the general pattern of what I'm doing.

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: I'll at least clean out the HandleRef around the pinned arrays in the initial work.

File: src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        [ActiveIssue(20884, TestPlatforms.AnyUnix)]
+        [ConditionalFact(Helpers.GdiplusIsAvailable)]
+        public void DrawBezier_Busy_ThrowsInvalidOperationException()
+        {
+            using (var image = new Bitmap(10, 10))
+            using (Graphics graphics = Graphics.FromImage(image))
+            using (var pen = new Pen(Color.Red))
+            {
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs]: This hdc ought to have a disposable wrapper perhaps

Comment (JeremyKuhne) [reply src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs]: The finally releases it, but it might be nice to add a helper.

**Comment by JeremyKuhne** [comment]
Did another update:

- Converted more usages
- Stopped wrapping the pointers
- Fully removed some of the helpers

Should be enough to see the pattern pretty clearly. There is a lot left to remove, but this is a good start I think. Significant cut in memory usage and should get at _least_ 10% better performance with methods that work with arrays. That gain will get bigger the larger the arrays are.

**Comment by JeremyKuhne** [comment]
On the upper end, transforming points is wayyy faster (26X).

Before:

```
   System.Drawing.Common.Performance.Tests.dll                          | Metric                                        | Unit  | Iterations |    Average | STDEV.S |        Min |        Max
  :-------------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| ----------:| -------:| ----------:| ----------:
   System.Drawing.Tests.Perf_Graphics_TransformPoints.Transform_Points  | Duration                                      | msec  |    100     |     94.878 |   5.830 |     91.804 |    140.345
   System.Drawing.Tests.Perf_Graphics_TransformPoints.Transform_Points  | Allocation Size on Benchmark Execution Thread | bytes |    100     | 1.248E+007 |   0.000 | 1.248E+007 | 1.248E+007
```

After:

```
   System.Drawing.Common.Performance.Tests.dll                          | Metric                                        | Unit  | Iterations | Average | STDEV.S |    Min |    Max
  :-------------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| -------:| -------:| ------:| ------:
   System.Drawing.Tests.Perf_Graphics_TransformPoints.Transform_Points  | Duration                                      | msec  |    100     |   3.584 |   1.129 |  3.181 | 13.747
   System.Drawing.Tests.Perf_Graphics_TransformPoints.Transform_Points  | Allocation Size on Benchmark Execution Thread | bytes |    100     |   0.000 |   0.000 |  0.000 |  0.000
```

We're riddled with this sort of overhead in our oldest UI stacks. I hope to change that. 😜

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: I find such changes that aren't necessary distracting from the rest of the PR. And the style guidelines don't say that this new version is preferred, so it's not actually fixing anything, just changing it to your preference; the next person who comes along to edit this file might change it back.

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > the style guidelines don't say that this new version is preferred

Perhaps they should so we don't rathole on nits...

>  it's not actually fixing anything, just changing it to your preference

The diff isn't hard to follow. I think I'm entitled to cleaning up the methods I'm changing here for style consistency that I prefer and that I personally think makes an improvement in readability.  I've made these methods concise & clean and made a _major_ improvement. I'll end up changing most of the file, so Jeremy's style will be predominant in this one file. _And_ if you look at the other changes I've made outside of this one, you'll see my style is predominant in the other files in this project. Shouldn't there be an argument for dll consistency as well?

> the next person who comes along to edit this file might change it back

If someone else comes in and makes these methods even better than I did (improves them functionally), more power to them- I think they've earned the right to make a minor style tweak in their favor.


Quite frankly this is really disheartening. I've made a big improvement on the way to even more big improvements and the key feedback I've gotten is a long debate over whether I should have removed an unnecessary set of brackets.

All that said, I do understand the intent of the rules and your concerns. I do think about the impact of what I do under those specific criteria (among others) and I try to exercise my best judgement in balancing trade-offs. I think I've chosen well here, or at least well enough that further arguments for alternatives are just niggling.

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > Quite frankly this is really disheartening. I've made a big improvement on the way to even more big improvements and the key feedback I've gotten is a long debate over whether I should have removed an unnecessary set of brackets.

I'm sorry it's disheartening. That's certainly not the intent. I provided feedback on everything I had to provide feedback on: this, validating the HandleRef usage, and questioning the interop. Your original change (prior to updating it after questions) was only a few lines of product src, so not a ton to comment on, and this style change was a large portion of the diff. You chose to push back on the feedback, from multiple people. When I've received such feedback, I've accepted it and moved on.

Comment (jkotas) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: @JeremyKuhne Feedback similar to what you got here was given number on number of occasions, by multiple different reviewers. We are just being consistent here. For example, here is my comment where I am asking Ashon to submit formatting changes separately from actual code changes: https://github.com/dotnet/coreclr/pull/16223#issuecomment-363641471

It is fine with me if you plan to make significant improvements in these files, want to claim their ownership and change them to your liking. Such style changes should be done in separate PR, and ideally the entire files should be changed to have consistent style, not just a few functions. Here are examples of PRs where component owners are changing the code to their liking  https://github.com/dotnet/corefx/pull/22471 or https://github.com/dotnet/coreclr/pull/17451

Comment (danmosemsft) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: @jeremykuhne I suggest moving formatting changes in a separate PR. You're right that this code could be better formatted but reviewers are asking not to do it in this change.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs [hunk 5]
```diff
@@ -183,46 +183,36 @@ public FillMode FillMode
...
-            int numPts = PointCount;
+            PathData pathData = new PathData()
+            {
+                Types = new byte[count],
+                Points = new PointF[count]
+            };
 
-            PathData pathData = new PathData() { Types = new byte[numPts] };
+            if (count == 0)
+                return pathData;
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: How common is this case?  If common, consider moving it earlier and doing:
```C#
if (count == 0)
{
    return new PathData() { Types = Array.Empty<byte>(), Points = Array.Empty<PointF>() };
}
```

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs [hunk 6]
```diff
@@ -183,46 +183,36 @@ public FillMode FillMode
...
-                    }
+                int status = SafeNativeMethods.Gdip.GdipGetPathData(new HandleRef(this, nativePath), &data);
 
-                    pathData.Points = SafeNativeMethods.Gdip.ConvertGPPOINTFArrayF(memoryPoints, numPts);
-                }
-                finally
+                if (status != SafeNativeMethods.Gdip.Ok)
                 {
-                    typesHandle.Free();
+                    throw SafeNativeMethods.Gdip.StatusException(status);
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: There's a helper for this:
```C#
SafeNativeMethods.Gdip.CheckStatus(status);
```
https://github.com/dotnet/corefx/blob/b5b75ee8969f2c67689eafe7b1ba73d0bddda019/src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs#L261-L267

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: Not common.

File: src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.Windows.cs]: > We are just being consistent here. For example, here is my comment where I am asking Ashon to submit formatting changes separately from actual code changes

And in that comment you explicitly said to restrict formatting changes to the areas where you're actually changing code. Which I agree with fully and make every attempt to do and I believe I've done in this case. (And I've pushed others to do.)

> It is fine with me if you plan to make significant improvements in these files, want to claim their ownership and change them to your liking. Such style changes should be done in separate PR, and ideally the entire files should be changed to have consistent style, not just a few functions.

I'm happy to do a follow up that makes a pass through the file(s) for consistency.

This caught me off guard. This is the _first_ time I've been tagged for changing the formatting of the code local to what I'm working on (honestly, it is just one line away from "real" changes). It really bothers me as it seems super nitpicky/dogmatic- did you really have a hard time understanding what changed? To me I can't see how it is in the least bit confusing- perhaps I just don't have the right perspective.

I'll put back the brackets. Hopefully my time explaining my perspective isn't wasted. I think we need to be careful about dogma and it worries me that we make things unnecessarily difficult.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: Yes, I plan to rationalize out the variety of helpers in a separate change.

**Comment by JeremyKuhne** [comment]
I've put back brackets around the if statements where I made no other change to said if statements.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/Matrix.cs [hunk 7]
```diff
@@ -261,134 +261,80 @@ public void Invert()
...
-        public void TransformPoints(PointF[] pts)
+        public unsafe void TransformPoints(PointF[] pts)
         {
             if (pts == null)
-                throw new ArgumentNullException("pts");
-            IntPtr buf = SafeNativeMethods.Gdip.ConvertPointToMemory(pts);
+                throw new ArgumentNullException(nameof(pts));
 
-            try
+            fixed(PointF* p = pts)
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/Matrix.cs]: Nit: space before (

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 8]
```diff
@@ -378,5 +378,8 @@
     <Reference Include="System" />
     <Reference Include="System.Drawing" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Include="System\Drawing\Internal\GpPathData.cs" />
```
Comment (safern) [inline src/System.Drawing.Common/src/System.Drawing.Common.csproj]: If this is a Windows specific could you move it up to the ItemGroup conditioned to TargetWindows L207? if we plan to use this in Unix code as well, could you please move it up the the common ItemGroup L29?

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Sure

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs [hunk 9]
```diff
@@ -1094,29 +1084,22 @@ public byte[] PathTypes
...
+        public unsafe PointF[] PathPoints
         {
             get
             {
-                int count = PointCount;
-                int size = Marshal.SizeOf(typeof(GPPOINTF));
-                IntPtr buf = Marshal.AllocHGlobal(checked(count * size));
-                try
+                PointF[] points = new PointF[PointCount];
+                fixed(PointF* p = points)
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: Nit: space before (

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/Matrix.cs [hunk 10]
```diff
@@ -261,134 +261,81 @@ public void Invert()
...
             if (pts == null)
-            {
-                throw new ArgumentNullException("pts");
-            }
+                throw new ArgumentNullException(nameof(pts));
 
-            IntPtr buf = SafeNativeMethods.Gdip.ConvertPointToMemory(pts);
-
-            try
+            fixed(Point* p = pts)
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/Matrix.cs]: Nit: space before (

File: src/System.Drawing.Common/tests/DrawingTest.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// See the LICENSE file in the project root for more information.
+
+using System.Drawing.Imaging;
+using System.IO;
+using Xunit;
+
+namespace System.Drawing.Tests
+{
+    public abstract class DrawingTest
```
Comment (safern) [inline src/System.Drawing.Common/tests/DrawingTest.cs]: Should we name this DrawingTestBase?

File: src/System.Drawing.Common/tests/Performance/Configurations.props [hunk 12]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <BuildConfigurations>
+      netstandard;
```
Comment (safern) [inline src/System.Drawing.Common/tests/Performance/Configurations.props]: Should we run the perf tests on netcoreapp only? The netstandard implementation is a turd assembly. In fact the regular tests only have a netcoreapp and desktop configurations

File: src/System.Drawing.Common/tests/DrawingTest.cs (hunk 11 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/tests/DrawingTest.cs]: I haven't used Base in other assemblies.

File: src/System.Drawing.Common/tests/Performance/Configurations.props (hunk 12 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/tests/Performance/Configurations.props]: Do we run them anywhere else? I was just copy/pasting from FileSystem. If we can/should specify netcoreapp I'm happy to change it.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPathIterator.cs [hunk 13]
```diff
@@ -169,73 +168,51 @@ public unsafe int Enumerate(ref PointF[] points, ref byte[] types)
...
-            byte[] typesLocal = new byte[count];
+            if (points.Length == 0)
+                return 0;
 
-            IntPtr memoryPts = Marshal.AllocHGlobal(checked(count * size));
-            try
+            fixed (PointF* p = points)
+            fixed (byte* t = types)
             {
-                int status = SafeNativeMethods.Gdip.GdipPathIterEnumerate(new HandleRef(this, nativeIter), out resultCount,
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPathIterator.cs]: Would be nice to rename nativeIter to nativeIterator, or something like this, perhaps not in the scope of this change.

File: src/System.Drawing.Common/tests/Performance/Configurations.props (hunk 12 above)
Comment (safern) [reply src/System.Drawing.Common/tests/Performance/Configurations.props]: I’m don’t think we yet do but if in the future we start it might cause pain to whoever brings the infra to do so, but I can live with this being NS

File: src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+                Point[] points =
+                {
+                    new Point(10, 10), new Point(20, 1), new Point(35, 5), new Point(50, 10),
+                    new Point(60, 15), new Point(65, 25), new Point(50, 30)
+                };
+
+                graphics.DrawBeziers(pen, points);
+                ValidateImageContent(image,
+                    PlatformDetection.IsWindows
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs]: Just curious, why is `IsWindows` needed if this test is already conditioned on GDI+ availability? What is the other platform?

Comment (JeremyKuhne) [reply src/System.Drawing.Common/tests/Graphics_DrawBezierTests.cs]: The check for GDI+ availability is also checking for the mono version of GDI+, which we use on Unix. The implementation isn't exactly the same and has some gaps.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPathIterator.cs (hunk 13 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPathIterator.cs]: I'm going to follow up this change with a style pass. I'll look at private/internal members.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs (hunk 9 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: D'oh, I searched and didn't click on this result. I'll get it in the next change.

File: src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs [hunk 15]
```diff
@@ -183,46 +183,36 @@ public FillMode FillMode
...
             {
-                GCHandle typesHandle = GCHandle.Alloc(pathData.Types, GCHandleType.Pinned);
-                try
+                GpPathData data = new GpPathData
                 {
-                    IntPtr typesPtr = typesHandle.AddrOfPinnedObject();
+                    Count = count,
+                    Points = p,
+                    Types = t
+                };
```
Comment (filipnavara) [inline src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: Doesn't this have to be fixed as well?

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Drawing2D/GraphicsPath.Windows.cs]: > or is it guaranteed to be on stack and hence not moving

Correct.  The language won't even let you use fixed here; it'll produce an error something about not letting you use fixed with an already fixed expression.
